npasswd-1.X/checkpasswd/checkpasswd.8
'\"
'\"	@(#)checkpasswd.8	1.1 5/18/89 (cc.utexas.edu) /tmp_mnt/usr/share/src/private/ut/share/bin/passwd/checkpasswd/SCCS/s.checkpasswd.8
'\"
.TH CHECKPASSWD 8
.SH NAME
checkpasswd \- check passwords
.SH SYNOPSIS
.B checkpasswd
[
.B \-\^c
config_file ] [
.B \-\^e
] [
.B \-\^o
] [
.B \-\^s
] [
.B \-\^u
user id ] [
.B \-\^V
] [
.B \-\^?
]
.SH DESCRIPTION
.I Checkpasswd
reads passwords from the standard input and subjects them to suitablity tests.
These tests are detailed below.
It purposely does
.B not
take a password argument since process
arguments are not secure on most UNIX systems - they can be printed by 
.IR ps (1).
.PP
.I Checkpasswd
is not meant as a replacement for 
.IR passwd (1)
and so does not suppress character echo while reading passwords.
The standard input can be redirected from a pipe or file, which facilitates
it being called from another program such as 
.IR passwd (1).
.PP
.I Checkpasswd
exits with the check status for the most recent password entered.
The check status is one of the following:
.nf
-1	A serious error occured.
0	The password passed all checks.
1	The password was a null string.
2	The password was \'too easy\' to guess.
3	The password was part of the users \'\fIfinger\fP\' information.
4	The password was found in a dictionary search.
5	The password contained an illegal character or sequence.
6	The password was too short.
.fi
.PP
In addition to the check status, an informative message is printed to
the standard output (unless supressed).
.PP
The options available are:
.TP
.B \-\^c
Take the next argument as the configuration file to read.
See the CONFIGURATION section for information on the configuration file.
.TP
.B \-\^e
Print the check status in numeric form as well as the informative message.
This is useful for programs which want to process the output from
.IR checkpasswd .
.TP
.B \-\^o
Test one password only and exit with check status.
.TP
.B \-\^s
Supress output of informative messages.
.TP
.B \-\^u
Take the next argument as the user id (in numeric form) to use for password
file data checking.
This option is restricted to the super-user.
.TP
.B \-\^V
Print version information.
.TP
.B \-\^?
Print help message.
.SH "CHECKS PERFORMED"
.PP
.I Checkpasswd
performs the following tests on each password:
.IP 1.
Length is checked to see if it is within bounds.
Passwords which are too short are rejected, while passwords that are
\'too long\' are passed but a warning message issued.
.IP 2.
The password is scanned for illegal characters and character combinations,
such as runs of the same character (e.g. \'aaa\').
Passwords that do not use upper and lower case alphabetics will be rejected
unless otherwise specified via the configuration file.
.IP 3.
The password is checked against various per-site information, such as the
host name.
.IP 4.
The password is checked against the
.IR passwd (5)
information for the user.
The password is compared against a number of permutations of this data.
.IP 5.
The password is checked against the user\'s \'finger\' information.
.IP 6.
The password is search for in a list of dictionaries and is rejected if it
is found in any of them.
.SH CONFIGURATION
.PP
Many of the criteria used by
.I checkpasswd
can be set via a configuration file.
The default configuration file is 
.BR /usr/adm/checkpasswd.cf .
.PP
Lines in the configuration file which start with '#' ignored.
Items in \'[ ]\' are optional, the default value is in \'{ }\'.
.sp
.nf
\fBdictionary\fP	/path/to/dictionary	[description of dictionary]
.fi
.RS
Specifies a dictionary to look in.
If there is a DBM data base version of the dictionary, that will be used,
else 
.IR egrep
will be used on the file.
There may be multiple
.B dictionary
directives in the configuration file, and they will be searched in
the order given.
A list of default dictionaries may be built into
.IR checkpasswd .
.br
The
.I makedict
program can be used to build DBM dictionaries, and
.I viewdict
used to review them.
.RE
.sp
.nf
\fBsinglecase\fP	yes | {no}
.fi
.RS
Specifies whether or not single-case passwords are legal.
The default is to reject single-case passwords.
.RE
.sp
.nf
\fBminlength\fP	N {5}	
.fi
.RS
Specifies the minimum password length to accept.
Passwords shorter that this will not be accepted.
.RE
.sp
.nf
\fBmaxlength\fP	N {8}
.fi
.RS
Specifies the maximum effective password length.
This does not affect acceptance of a password.
The 
.IR crypt (3)
routine usually encrypts only the first 8 characters of a string due to the
algorithim used.
If the password is longer than this, a warning message is printed to inform the
user that not the entire password will be used.
.RE
.sp
.nf
\fBprintonly\fP	yes | {no}
.fi
.RS
Specifies that only printable ASCII characters are to be allowed in passwords.
Use of control characters should be encouraged but may cause problems on
some systems.
.RE
.sp
.nf
\fBbadchars\fP	"list-of-characters"
.br
\fBbadchars\fP	+"list-of-characters"
.fi
.RS
Specifies a list of characters which are not allowed in passwords.
The first form replaces the illegal character table.
The second form adds to the illegal character table.
Characters may be given in the following forms:
.br
.ti +.25i
C special character escapes (e.g. \\\|t )
.br
.ti +.25i
\\\|0[0\-7] (e.g. \\\|013)
.br
.ti +.25i
\\\|0x[0\-9\| \|a\-f] (e.g. \\\|0xa)
.br
.ti +.25i
^[a\-z\|,\|A\-Z\|,\|[\|,\|\\\|,\|]\|,\|^\|,\|-] (e.g.,^c)
.br
.sp
The default content of the illegal characters table is a collection
of terminal special characters.
.RE
.in 0
.SH SEE ALSO
makedict, viewdict, passwd(1)
.SH AUTHOR
Clyde Hoover
.br
Computation Center
.br
The University of Texas at Austin
.br
Austin, Texas
.br
clyde@emx.utexas.edu, uunet!cs.utexas.edu!ut-emx!clyde
.PP


v2.1/NEWS
Latest npasswd news	$Date: 1999/02/01 00:00:00 $

Patch 05	2/1/99	A passel of little bug fixes
	* Added copyright messages to more modules.
	* A bit of code format cleanup.
	* Do not enable nsswitch support unless the "nsswitch.h" include
	  file is present.  This means that nsswitch.conf will ONLY be
	  supported on Solaris 2.
	  IRIX 6.5 and Redhat Linux 4.2 have /etc/nsswitch.conf but totally
	  different API for support.

	* Properly punt the '-r' option on SunOS 5.
	* Fixed minor Makefile things throughout the distribution.
	* Added the option "-XA" to always prompt for the old password.
	* Fixed lexical password check to properly recoginize telephone numbers.
	* Enhanced debug mode in checkpassword program.
	* Use an internal pager to present help files.
	* Reorganize terminal handling code.
	* Fix -XI option to work right when stdin is a pipe
	* Fix Configure password file list input loop
	* Re-activate some Cracklib rules to be more anal about password
	  checking.

Patch 04	10/15/98	Lots of bug fixes, some feature enhacements
	* Substantial rework of history_admin:
		* Add 'merge' function (see manual page)
		* Add '-i' command line option 
		* Allow '-m' with default database
		* Fix some more annoying bugs
	* Modifications to the password history test suite to use
	  and verify the above changes
	* Clarify some issues with password history depth
		* History depth now means "how many *old* passwords matter"
		* Always retain the more recent password, even if too old
	* Allow "@" to mean the default database name in history
	  configuration directives (and the history_admin '-f' option)
	* Fix the "passwd.MaxRepeat" config directive
	* Fix the "ShadowTolerance" config directive
	* Fix the "PasswdTolerance" config directive
	* Fix the "PasswdMatchTries" config directive
	* Add build-time option "SHORT_FILE_WARN" to control behavior
	  when passwd/shadow temp files come up short
	* Support the '-r' option under SunOS 5
	* Allow change of local passwd even if NIS client (unless SunOS 4)
	* More run-time trace messages in various places

Patch 03	9/24/98
	* Fixes the "Cannot talk to NIS server" problem under Solaris 2. 

Patch 02	9/22/98
	* Fixes the build problems on Digital UNIX 4.0d.

Patch 01	9/18/98
	* Fixes build problems when NIS support is not enabled.
	* Fixes a problem with the configuration directive "PrintOnly", which
	  was renamed "PrintableOnly" during the last development cycle but
	  the change did not get propigated everywhere.

Platform status:

AIX 4.1		Builds, installs, changes local passwords
SunOS 4.1.4	Builds, installs, changes local and NIS passwords 
SunOS 5.5	Builds, installs, changes local and NIS passwords 
IRIX 6.1	Builds, installs, changes local passwords 
Digital UNIX 4.0	Builds, installs



V2.0/contrib/maybe/bloom/checkpasswd.c

/* --------------------------------------------------------------------  */
/*                                                                       */
/*                         Author: Clyde Hoover                          */
/*                          Computation Center                           */
/*                   The University of Texas at Austin                   */
/*                          Austin, Texas 78712                          */
/*                         clyde@emx.utexas.edu                          */
/*                   uunet!cs.utexas.edu!ut-emx!clyde                    */
/*                                                                       */
/*This code may be distributed freely, provided this notice is retained. */
/*                                                                       */
/* --------------------------------------------------------------------  */
/*
 *	checkpasswd.c - Password check driver and data initialization
 */

#ifndef lint
static char sccsid[] = "@(#)checkpasswd.c	1.1 5/18/89 (cc.utexas.edu)";
#endif

#include "checkpasswd.h"

/*
 *	Table of password check parameters
 *	May be modified via the configuration file
 */
int	single_case =	0,		/* Single-case pwds ok */
	print_only =	0,		/* Printable ASCII chars only */
	run_length =	3,		/* How long chars runs can be */
	min_length =	5,		/* Minimum length */
	max_length =	8;		/* Maximum effective length */

/*
 *	Control characters best avoided - commonly-used terminal controls.
 *	Add characters here or replace entire contents via the
 *	configuration file.
 */
#define	ctrl(d)	('d' & 037)

char	illegalcc[sizeof_illegalcc] = {
	ctrl(c),	/* Interrupt character */
	ctrl(d),	/* UNIX end-of-file */
	ctrl(h),	/* Backspace */
/* 	ctrl(i), */
	ctrl(j),	/* Newline */
	ctrl(m),	/* Carriage return */
	ctrl(o),	/* Flush output */
	ctrl(r),	/* Retype pending input */
	ctrl(s),	/* Suspend output */
	ctrl(q),	/* Resume output */
	ctrl(y),	/* Suspend program deferred */
	ctrl(z),	/* Suspend program immediate */
	ctrl(\\),	/* Quit signal */
	ctrl([),	/* escape - may do strange things to ttys if echoed */
	ctrl(]),	/* UNIX telnet escape */
	'0177',		/* rubout */
	0
};

/*
 *	The 'pwck_*' routines all use the PWCK_* return
 *	codes, which are then propigated up to the caller of checkpassword().
 *
 *	All pwck_* routines in the table below are called thusly:
 *		pwck_*(password, userid, mesg)
 *			password = plaintext password string to test.
 *			userid = the user id which wants to use <password>.
 *			mesg = buffer to place long explanation into
 *
 *	If more checks are desired, add the functions to the tables below.
 */
extern int
	pwck_lexical(),
	pwck_local(),
	pwck_passwd(),
	pwck_finger(),
	pwck_dictionary();

typedef	int	(*function)();

function checkprocs[] = {
	pwck_lexical,
	pwck_local,
	pwck_passwd,
	pwck_finger,
	pwck_dictionary,
	0
};

/*
 *	checkpassword - Password candidate sanity checker.
 *
 *	Arguments;
 *		password = plain text password string to check.
 *		userid = the uid whom the password is for, -1 to disable.
 *
 *	Returns:
 *		PWCK_* values (see checkpasswd.h)
 */
checkpassword(password, userid, mesg)
char	*password;		/* Plaintext of password to check */
int	userid;			/* The user this is for */
char	*mesg;			/* Where to stash explanation message */
{
	int		rcode;		/* General purpose scratch */
	function	*checkfunc;	/* Check function pointer */

	if (password == 0 || *password == 0)
		return(PWCK_NULL);		/* Null password */

	mesg[0] = 0;
	for (checkfunc = checkprocs; *checkfunc; checkfunc++) {
		if ((rcode = (*checkfunc)(password, userid, mesg)) != PWCK_OK)
			return(rcode);
	}
	return(PWCK_OK);
}
/*	End checkpasswd.c */
